File: 53.maximum-subarray.py
# @leet start
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        
        global_max = nums[0]
        local_max = nums[0]

        for i in range(1, len(nums)):
            val = nums[i]
            if local_max < 0:
                local_max = max(val, local_max)
            else:
                local_max += val
            global_max = max(global_max, local_max)
        return global_max
# @leet end

File: 53.maximum-subarray.cpp
// @leet start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_sum = nums[0];
        int curr_sum = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            if (curr_sum < 0) {
                curr_sum = max(curr_sum, nums[i]);
            } else {
                curr_sum += nums[i];
            }
            max_sum = max(curr_sum, max_sum);
        }
        
        return max_sum;
    }
};
// @leet end
